File: BusinessObjects/Entities/ElementCell.cs
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations.Schema;
using System.Linq;

namespace forCrowd.Backbone.BusinessObjects.Entities
{
    public class ElementCell : BaseEntity
    {
        public ElementCell()
        {
            UserElementCellSet = new HashSet<UserElementCell>();
        }

        public int Id { get; set; }

        [Index("UX_ElementCell_ElementFieldId_ElementItemId", 1, IsUnique = true)]
        public int ElementFieldId { get; set; }

        [Index("UX_ElementCell_ElementFieldId_ElementItemId", 2, IsUnique = true)]
        public int ElementItemId { get; set; }

        public string StringValue { get => stringValue; set => stringValue = value?.Trim(); }

        public decimal DecimalValueTotal { get; set; }
        public int DecimalValueCount { get; set; }

        /// <summary>
        /// In case this cell's field type is Element, this is the selected item for this cell.
        /// Other values are stored on UserElementCell, but since this one has FK, it's directly set on ElementCell.
        /// </summary>
        public int? SelectedElementItemId { get; set; }
        public ElementItem ElementItem { get; set; }
        public ElementField ElementField { get; set; }
        public ElementItem SelectedElementItem { get; set; }
        public ICollection<UserElementCell> UserElementCellSet { get; set; }
        public UserElementCell UserElementCell => UserElementCellSet.SingleOrDefault();

        string stringValue;

        public void SetValue(ElementItem value)
        {
            SetValueHelper(ElementFieldDataType.Element);
            SelectedElementItem = value;
        }

        public ElementCell SetValue(string value)
        {
            SetValueHelper(ElementFieldDataType.String);
            StringValue = value;
            return this;
        }

        private void SetValueHelper(ElementFieldDataType valueType)
        {
            // Validations

            // a. Field and value types have to match
            var fieldType = (ElementFieldDataType)ElementField.DataType;

            if (fieldType != valueType)
            {
                throw new System.InvalidOperationException(
                    $"Invalid value, field and value types don't match - Field type: {fieldType}, Value type: {valueType}");
            }

            // Clear, if FixedValue
            if (ElementField.UseFixedValue)
                ClearFixedValues();
        }

        private void ClearFixedValues()
        {
            //StringValue = null;
            //DecimalValue = null;
            // TODO Do we need to set both?
            SelectedElementItemId = null;
            SelectedElementItem = null;
        }
    }
}

File: AngularClient/projects/forcrowd-backbone/src/lib/entities/element-cell.ts
import { EntityBase } from "./entity-base";
import { ElementField } from "./element-field";
import { ElementItem } from "./element-item";
import { UserElementCell } from "./user-element-cell";

export class ElementCell extends EntityBase {

  // Public - Server-side
  Id = 0;
  ElementField: ElementField;
  ElementItem: ElementItem;
  DecimalValueTotal = 0;
  DecimalValueCount = 0;
  SelectedElementItem: ElementItem;
  UserElementCellSet: UserElementCell[];

  get StringValue(): string {
    return this.fields.stringValue;
  }
  set StringValue(value: string) {
    this.fields.stringValue = value ? value.trim() : null;
  }

  // Client
  otherUsersDecimalValueTotal = 0;
  otherUsersDecimalValueCount = 0;

  // Client-side
  private fields: {
    stringValue: string,
  } = {
      stringValue: null,
    };

  initialize(): boolean {
    if (this.initialized) return;

    super.initialize();

    // Other users'
    this.otherUsersDecimalValueTotal = this.DecimalValueTotal;
    this.otherUsersDecimalValueCount = this.DecimalValueCount;

    // Exclude current user's
    if (this.UserElementCellSet[0]) {
      this.otherUsersDecimalValueTotal -= this.UserElementCellSet[0].DecimalValue;
      this.otherUsersDecimalValueCount -= 1;
    }

    // User cells
    this.UserElementCellSet.forEach(userCell => {
      userCell.initialize();
    });
  }
}
